### PR TITLE
ArrayList Optimizations and Byte deserialization optimizations

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
* Tune some usage of ArrayLists and Optional
* Avoid decoding strings unnecessarily

Before optimization:
Total time: PT1.602132S (Cold JVM?)
Total time: PT1.25578S
Total time: PT1.187478S
Total time: PT1.203511S
Total time: PT1.176007S
Total time: PT1.260642S
Total time: PT1.255856S
Total time: PT1.310806S
Total time: PT1.442327S
Total time: PT1.310608S
Total time: PT1.222053S
Total time: PT1.220536S

After:
Duration: PT1.274681S (Cold JVM?)
Duration: PT0.927487S
Duration: PT0.869735S
Duration: PT0.920447S
Duration: PT0.882247S
Duration: PT0.871624S
Duration: PT0.883032S
Duration: PT0.884062S
Duration: PT0.88226S
Duration: PT0.897061S
Duration: PT0.907535S

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -96,7 +96,7 @@ repositories {
     }
   }
 }
-val fgputil = "2.7.4-jakarta" // FgpUtil version
+val fgputil = "2.7.5-jakarta-SNAPSHOT" // FgpUtil version
```
Comment (ryanrdoherty) [inline build.gradle.kts]: Will want to fix this up before merging.

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java [hunk 2]
```diff
@@ -168,9 +170,14 @@ public static void produceTabularSubsetFromFile(Study study, Entity outputEntity
 
     final DataFlowTreeReducer driver = new DataFlowTreeReducer(idIndexEntityConverter, binaryValuesStreamer);
     try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(outputStream))) {
-      final ResultConsumer resultConsumer = formatter.getFormatter(writer);
+      final TabularResponses.BinaryResultConsumer resultConsumer = formatter.getFormatter(outputStream);
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java]: Probably don't want to mix and match the OutputStream and the BufferedWriter.  Maybe pick one for the try-with-resources and only use that?

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java [hunk 3]
```diff
@@ -168,9 +170,14 @@ public static void produceTabularSubsetFromFile(Study study, Entity outputEntity
...
 
-      resultConsumer.consumeRow(usePrettyHeader ? getTabularPrettyHeaders(outputEntity, outputVariables) : outputColumns);
+      resultConsumer.consumeRow(usePrettyHeader ? getTabularPrettyHeaders(outputEntity, outputVariables).stream()
+          .map(s -> s.getBytes(StandardCharsets.UTF_8))
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java:173]: Not quite following- is the formatter still adding the tab as UTF-8 encoded bytes between the byte[] objects passed in?  Wondering if lose anything by not buffering here.  I know the numbers are better with this but could we reintroduce buffering to improve even further?

File: src/main/java/org/veupathdb/service/eda/ss/model/tabular/TabularResponses.java [hunk 4]
```diff
@@ -16,21 +19,27 @@ public class TabularResponses {
   private TabularResponses(){}
 
   public enum Type {
-    TABULAR(TABULAR_FORMATTER, MimeTypes.TEXT_TABULAR),
-    JSON(JSON_FORMATTER, MediaType.APPLICATION_JSON);
+    TABULAR(TABULAR_FORMATTER, MimeTypes.TEXT_TABULAR, TABULAR_BINARY_FORMATTER),
+    JSON(JSON_FORMATTER, MediaType.APPLICATION_JSON, TABULAR_BINARY_FORMATTER);
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/eda/ss/model/tabular/TabularResponses.java]: Need to implement JSON version here.  Also (nit) would prefer the arg order be (formatter, binary-formatter, mime-type), to put those two together.

**Review by ryanrdoherty — CHANGES_REQUESTED** [review]
I wonder if you should add a JSON request or two to your regression comparison test script?

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java (hunk 2 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java]: Yeah, oops, that was an unused variable. Cleaning things up a bit in the next revision. Added the buffering back and removed the writer altogether for the file-based method.

File: src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java (hunk 3 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/db/FilteredResultFactory.java:173]: Yeah there are some places where we have the encoded bytes already but this is a place where we need to encode the string and a `Writer` is a more natural interface.

The buffering being removed was an oversight by me. I'm adding it back in revision 2. I'll do some testing to see how it looks.

File: src/main/java/org/veupathdb/service/eda/ss/model/tabular/TabularResponses.java (hunk 4 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/service/eda/ss/model/tabular/TabularResponses.java]: I thought I had this locally, but I must not have committed. Adding back now

**Pull request merged**